src/libraries/System.IO.Compression/src/System/IO/Compression/ZipHelper.cs

```diff
@@ -209,24 +209,36 @@ internal static byte[] GetEncodedTruncatedBytesFromString(string? text, Encoding
 
             encoding ??= GetEncoding(text);
             isUTF8 = encoding.CodePage == 65001;
-            byte[] bytes = encoding.GetBytes(text);
 
-            return GetEncodedTruncatedBytes(bytes, encoding, maxBytes);
-        }
+            if (maxBytes == 0) // No truncation
+            {
+                return encoding.GetBytes(text);
+            }
 
-        internal static byte[] GetEncodedTruncatedBytes(byte[] bytes, Encoding encoding, int maxBytes)
-        {
-            if (maxBytes > 0 && bytes.Length > maxBytes)
+            byte[] bytes;
+            if (isUTF8)
             {
-                int bytesPerChar = encoding.GetMaxByteCount(1);
+                int totalCodePoints = 0;
+                foreach (Rune rune in text.EnumerateRunes())
+                {
+                    if (totalCodePoints + rune.Utf8SequenceLength > maxBytes)
+                    {
+                        break;
+                    }
+                    totalCodePoints += rune.Utf8SequenceLength;
+                }
 
-                int encodedCharsThatFit = maxBytes / bytesPerChar;
-                int totalBytesToTruncate = encodedCharsThatFit * bytesPerChar;
+                bytes = encoding.GetBytes(text);
 
-                return bytes[0..totalBytesToTruncate];
+                Debug.Assert(totalCodePoints > 0);
+                Debug.Assert(totalCodePoints <= bytes.Length);
+
+                return bytes[0..totalCodePoints];
             }
 
-            return bytes;
+
+            bytes = encoding.GetBytes(text);
+            return maxBytes < bytes.Length ? bytes[0..maxBytes] : bytes;
```

Comment:
`bytes[0..maxBytes]` is an allocation of it's own, isn't it? If so, consider implementing ArrayPool.

Reply:
This was discussed before. We don't expect either the `Comment` setter or the `EntryName` setter to be called in a loop. That would be weird.

Reply:
it is being called in a foreach loop in your own tests. 
https://github.com/dotnet/runtime/blob/3444631603caf724abaa2c6bdf2366e76b7d23b4/src/libraries/System.IO.Compression/tests/ZipArchive/zip_UpdateTests.Comments.cs#L67 

I don't know if enumeration of entries should care or not about one allocation, but it can definitely be improved to my suggestion.

Reply:
I opened issue https://github.com/dotnet/runtime/issues/64767 to follow up on this.